renderer/backend/DrawBatchExecutor.cpp

```diff
@@ -0,0 +1,237 @@
+#include "precompiled.h"
+#include "DrawBatchExecutor.h"
+
+#include "../glsl.h"
+#include "../qgl.h"
+
+const uint DrawBatchExecutor::DEFAULT_UBO_INDEX;
+const uint DrawBatchExecutor::MAX_SHADER_PARAMS_SIZE;
+
+namespace {
+	const uint MAX_DRAWS_PER_FRAME = 8192;
+
+	struct DrawElementsIndirectCommand {
+		uint count;
+		uint instanceCount;
+		uint firstIndex;
+		uint baseVertex;
+		uint baseInstance;
+	};
+}
+
+idCVarBool r_useMultiDrawIndirect("r_useMultiDrawIndirect", "1", CVAR_RENDERER|CVAR_BOOL, "Batch draw calls in multidraw commands if available");
+
+void DrawBatchExecutor::Init() {
+	GLint uboAlignment;
+	qglGetIntegerv( GL_UNIFORM_BUFFER_OFFSET_ALIGNMENT, &uboAlignment );
+	qglGetIntegerv(GL_MAX_UNIFORM_BLOCK_SIZE, &maxUniformBlockSize);
+
+	uint shaderParamsBufferSize = MAX_DRAWS_PER_FRAME * MAX_SHADER_PARAMS_SIZE;
+	shaderParamsBuffer.Init( GL_UNIFORM_BUFFER, shaderParamsBufferSize, uboAlignment );
+
+	if (GLAD_GL_ARB_multi_draw_indirect) {
+		InitDrawIdBuffer();	
+		uint drawCommandBufferSize = MAX_DRAWS_PER_FRAME * sizeof(DrawElementsIndirectCommand);
+		drawCommandBuffer.Init( GL_DRAW_INDIRECT_BUFFER, drawCommandBufferSize, 16 );
+		drawCommandBuffer.Bind();
+	}
+}
+
+void DrawBatchExecutor::Destroy() {
+	shaderParamsBuffer.Destroy();
+	qglDeleteBuffers(1, &drawIdBuffer);
+	drawIdBuffer = 0;
+	drawCommandBuffer.Destroy();
+}
+
+void DrawBatchExecutor::ExecuteDrawVertBatch( int numDrawSurfs, GLuint uboIndex ) {
+	assert( numDrawSurfs <= maxBatchSize );	
+	maxBatchSize = 0;
+
+	byte *shaderParamsContents = shaderParamsBuffer.CurrentWriteLocation();
+	uint shaderParamsCommitSize = numDrawSurfs * shaderParamsSize;
+	shaderParamsBuffer.Commit( shaderParamsCommitSize );
+	shaderParamsBuffer.BindRange( uboIndex, shaderParamsContents, shaderParamsCommitSize );
+
+	vertexCache.BindVertex();
+	vertexCache.BindIndex();
+	if ( ShouldUseMultiDraw() ) {
+		DrawVertsMultiDraw( numDrawSurfs );
+	} else {
+		DrawVertsSingleDraws( numDrawSurfs );
+	}
+
+	if ( r_showPrimitives.GetBool() && !backEnd.viewDef->IsLightGem() && backEnd.viewDef->viewEntitys ) {
+		for ( int i = 0; i < numDrawSurfs; ++i ) {
+			const drawSurf_t *surf = drawSurfs[i];
+			backEnd.pc.c_drawIndexes += surf->numIndexes;
+			if ( surf->frontendGeo ) {
+				backEnd.pc.c_drawVertexes += surf->frontendGeo->numVerts;
+			}
+			backEnd.pc.c_vboIndexes += surf->numIndexes;
+		}
+		backEnd.pc.c_drawElements += ShouldUseMultiDraw() ? 1 : numDrawSurfs;
+	}
+}
+
+void DrawBatchExecutor::ExecuteShadowVertBatch( int numDrawSurfs, GLuint uboIndex ) {
+	assert( numDrawSurfs <= maxBatchSize );	
+	maxBatchSize = 0;
+
+	byte *shaderParamsContents = shaderParamsBuffer.CurrentWriteLocation();
+	uint shaderParamsCommitSize = numDrawSurfs * shaderParamsSize;
+	shaderParamsBuffer.Commit( shaderParamsCommitSize );
+	shaderParamsBuffer.BindRange( uboIndex, shaderParamsContents, shaderParamsCommitSize );
+
+	vertexCache.BindVertex( ATTRIB_SHADOW );
```

Comment:
Why do you prefer having all methods duplicated for shadow?

It seems that `ATTRIB_SHADOW` and `sizeof(shadowCache_t)` are the only differences.

Reply:
I don't prefer it; I guess the methods ended up more similarly than I originally anticipated. Will try to unify.